### ISSUE TITLE
Standardize version numbers reported by builds

### ISSUE THREAD

**Issue by rfratto** [grafana/agent#6100]
### Background

Today, version numbers reported by the binary are inconsistent. They are either:

* `<BRANCH>-<SHA>` for non-tagged builds.
* `<TAG>` (i.e., `v0.39.0-rc.0`) for release builds. 
* `VERSION` (i.e., `0.39.0`) for release builds produced by Homebrew. 

This makes it difficult to track down what versions of agents are in use, and makes it difficult for configuration management tooling to make decisions based on non-release builds. (See #5384 for the first time this was suggested). 

I would like to standardize on reported versions to:

* Enable remote configuration to make consistent decisions based on the version of running agents. 
* Make it easier to track which versions are in use when deploying non-release builds at scale.
* Standardize on version strings regardless of where the binary was built (aligning builds produced by us and Homebrew)

### Proposal

I propose modifying our build process such that the Grafana Agent binary always reports valid [semver](https://semver.org/) versions, prefixed by a `v`:

* Release candidates: `vMAJOR.MINOR.0-rc.0` 
* Stable releases: `vMAJOR.MINOR.0` 
* Patch releases: `vMAJOR.MINOR.PATCH` 
* Non-release builds: `vMAJOR.MINOR.0-devel+SHA`, where SHA is a seven byte representation of the commit SHA. 

There are a few things to note here:

* This makes the `v` prefix (as in `v0.39.0`) always included; consistently being used in branch, tag, and versions. 
  * If the version given to the agent on build does not include the `v` prefix, it is injected into the reported version. This will fix the issue where Homebrew builds are the only builds that do not include a `v`.  
* After a release is published, non-release builds should point to the _next_ release version. That is to say, if `v0.39.0` was just released, commits to main should be versioned as `v0.40.0-devel+SHA`. This rule aligns with semantic versioning rules around prereleases to give it precedence over the previously published release. 

The consistent versioning is only guaranteed for builds produced via `make`; due to our dependency on specific build flags to inject version information, builds produced manually may not report the correct version.

As an artifact of this change, we must have a file which represents the `MAJOR.MINOR.PATCH` version to include in the binary. After a new release branch is produced, this file must be updated in main to the next release version, as detailed above. 

To summarize all the places where versions will be used:

* As reported by the Grafana Agent binary: `vVERSION` (with the `v` prefix)
* As named in the release branch: `release-vMAJOR.MINOR` (with the `v` prefix)
* As named in a Git tag: `vVERSION` (with the `v` prefix)

For prior work on this proposal, see #5384 and #5473.

**Comment by captncraig** [comment]
I completely agree except one point:

> After a release is published, non-release builds should point to the next release version.

My intuition would say that the development version should reference the _previous_ release. I feel like if somebody has an issue or something and they say they are running `v0.39.0-devel+a213ad`, my expectation is that they have all the features that went into 0.39, but if 0.39 is still in progress, that is not the case. They are actually v0.38.x plus some unknown amount of features on top. It seems minor when 38 and 39 are relatively recent, but into the future when 0.39 is associated with some set of big features, I would expect any build labeled 0.39.xxxx would include them.

**Comment by rfratto** [comment]
> My intuition would say that the development version should reference the previous release.

If you read the semantic versioning specification, this doesn't make sense: `devel` here indicates a prerelease version, which means that `v0.39.0` is _always_ considered newer than `v0.39.0-devel+SHA`, no matter what. Semantic versioning also specifies that there is no ordering or compatibility constraints between prerelease versions, so there's no way to differentiate between `v0.39.0-devel+deadbeef` and `v0.39.0-devel+deadd00d`. 

Here's the timeline as I see it:

* Today, the stable release is v0.39.0. 
* We are currently working on the release cycle for v0.40. Builds from main should start to report `0.40.0-devel+SHA`. 
  * There is no guarantee for what is available in 0.40.0-devel+SHA commit-by-commit; breaking changes may happen, it may be prone to bugs, etc. It's a pre-release build. 
* Once the `release-v0.40` branch is created, the main branch becomes the development branch for v0.41, and builds start reporting `0.41.0-devel+SHA`. 

Given the ordering constraint within the semver spec, referencing the previous release doesn't make sense to me: it doesn't make sense for the main branch to report `0.39.0-devel+SHA`, because 0.39.0 has already been released.

**Comment by mattdurham** [comment]
What does the time between an RC and actual release look like? My assumption that if we release 0.39.0-rc.0 that main is 0.40.0-devel?

**Comment by rfratto** [comment]
> What does the time between an RC and actual release look like? My assumption that if we release 0.39.0-rc.0 that main is 0.40.0-devel?

Slightly earlier than that; the second the `release-v0.39` branch is created, then main becomes `0.40.0-devel`, and all further development specific to v0.39 belongs in the release-v0.39 branch.

**Comment by captncraig** [comment]
But then future work off of the v0.39 branch (cherry picks and such) will be `v0.39.1-devel`? 

If this is all how semver is supposed to work, I guess I can get my head around it. I don't like the status quo at all, so let's do it.

**Comment by rfratto** [comment]
> But then future work off of the v0.39 branch (cherry picks and such) will be v0.39.1-devel?

The signal for when the version changes within a release branch is a little different, but yes: after 0.39.0 (the stable release) is published, then future work is 0.39.1-devel+SHA. The key indicator here is that the `devel` prerelease identifier is used to indicate what version _will_ be produced. (You can also see this same approach in use in some Go module versions if you ask Go to import a commit instead of a tag)

**Comment by rfratto** [comment]
I have updated this proposal so that _all_ version strings include the `v` prefix, including branch names, tag names, and binary versions. The proposal now includes a section noting that the `v` prefix will be injected if the binary is built with a valid semver version but no `v` prefix.

**Comment by rfratto** [comment]
This has been implemented in grafana/alloy#44. Closing!
